### PR TITLE
Operator simplification #1: Class level simplify methods

### PR DESCRIPTION
**Pull request by AlbertMitjans**
**Context:** Add `simplify` method to reduce the depth of nested operators.

**Description of the Change:**
- Add `arithmetic_depth` property and `simplify` method to `Operator`, `Sum`, `Pow`, `SProd` and `Adjoint` classes.
- Add tests.

**Benefits:** 
Users can now make use of the `simplify` method to reduce the depth of nested operators. 

Example:
```python
import pennylane as qml

sum_op = qml.ops.Sum(qml.PauliX(0), qml.PauliX(0))
for _ in range(10):
  sum_op += qml.PauliX(0)
print(sum_op.arithmetic_depth)
print(sum_op.simplify(depth=3).arithmetic_depth)
print(sum_op.simplify().arithmetic_depth)
```
Prints:
```
11
8
1
```
Another example:
```python
import pennylane as qml
adj_op = qml.ops.Adjoint(qml.ops.Adjoint(qml.ops.Sum(qml.PauliX(0), qml.ops.Sum(qml.PauliX(0), qml.PauliX(0)))))
simplified_op = adj_op.simplify()
qml.equal(op1=simplified_op, op2=qml.ops.Sum(qml.PauliX(0), qml.PauliX(0), qml.PauliX(0)))
```
Returns `True`

### CODE REVIEW

File: pennylane/operation.py
Comment (AlbertMitjans) [reply pennylane/operation.py]: Changed name to `arithmetic_depth` and made it a property in commit 47af2e61974cc24ef8759bc1e683d3dd21523c61.

**Comment by AlbertMitjans** [comment]
[sc-22449]

File: pennylane/ops/op_math/adjoint_class.py
Comment (albi3ro) [inline pennylane/ops/op_math/adjoint_class.py:246]:
```suggestion
        if isinstance(self.base, Adjoint):  # Adj(Adj(A)) = A
```
The `if depth==0` takes care of the second check.
If `depth==1`, then the base can't be an `Adjoint`.

File: pennylane/ops/op_math/adjoint_class.py
Comment (albi3ro) [inline pennylane/ops/op_math/adjoint_class.py:245]: That can be taken care of once we add the `Prod` class.

File: pennylane/ops/op_math/adjoint_class.py
Comment (albi3ro) [inline pennylane/ops/op_math/adjoint_class.py:243]: Since `if depth==0: return self` is so concise and readable, I think adding a decorator will make the interface more complicated than the code.

It will be harder for developers to remember the syntax for the decorator and track down its definition than to simply remember the code, then it would hurt more than it would help.

File: pennylane/ops/op_math/adjoint_class.py
Comment (albi3ro) [inline pennylane/ops/op_math/adjoint_class.py:249]:
```suggestion
            simplified_sum = self.base.simplify(depth=depth-1)
```

File: pennylane/ops/op_math/pow_class.py
Comment (albi3ro) [inline pennylane/ops/op_math/pow_class.py:253]: Remember that `Identity` can only accept a single wire at this point in time. 
```suggestion
            return Identity(wires=self.wires[0], id=self.id)
```

Might be a good temporary solution.

File: pennylane/ops/op_math/sprod.py
Comment (albi3ro) [inline pennylane/ops/op_math/sprod.py:211]:
```suggestion
            if scalar == 1:
                return self.base.base.simplify(depth=depth - 1)
```

**Review by albi3ro — COMMENTED** [review]
I think its worth thinking about whether we want `depth` to mean:

1) Number of times we run the the `simplify` method
2) Maximum decrease in `arithmetic_depth`

I think I prefer the first option.  Then we would only decrease `depth` by one when simplifying the next layer of operations.

If we only decrease by one, we don't have to worry about skipping from `1` to `-1`, bypassing zero.

File: pennylane/ops/op_math/adjoint_class.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/adjoint_class.py]: Should we add the `-1` here? If the depth is defined as the maximum number of stacked operators, when we do `Adj(Sum(A, B)) --> Sum(Adj(A), Adj(B))` we are not reducing the depth. That is why I didn't add the `-1`.

By adding the `-1` here then we would have inconsistencies such as:
```python
>>> import pennylane as qml
>>> adj_op = qml.adjoint(qml.op_sum(qml.op_sum(qml.PauliX(0), qml.PauliX(0)), qml.PauliX(0)))
>>> adj_op.arithmetic_depth
3
>>> sum_op = adj_op.simplify(depth=1)
>>> sum_op.arithmetic_depth
3
```

**Comment by AlbertMitjans** [comment]
> I think its worth thinking about whether we want `depth` to mean:
> 
> 1. Number of times we run the the `simplify` method
> 2. Maximum decrease in `arithmetic_depth`
> 
> I think I prefer the first option. Then we would only decrease `depth` by one when simplifying the next layer of operations.
> 
> If we only decrease by one, we don't have to worry about skipping from `1` to `-1`, bypassing zero.

We can surely do it like the first option.

**Comment by albi3ro** [comment]
Make sure this will work with queuing. For example:

```
with qml.tape.QuantumTape() as tape:
    op = qml.adjoint(qml.adjoint(qml.RX(1.2, wires=0)))
    qml.simplify(op)
```
In this case, `tape.circuit` should just have the original `RX` gate. This might get a bit more complicated.

File: pennylane/ops/functions/equal.py
Comment (albi3ro) [inline pennylane/ops/functions/equal.py:22]: We can probably move the changes to the equal function to its own PR.

File: pennylane/ops/op_math/adjoint_class.py
Comment (albi3ro) [inline pennylane/ops/op_math/adjoint_class.py:245]:
```suggestion
        if isinstance(self.base, Adjoint) and (depth > 1):  # Adj(Adj(A)) = A
```
The nice thing about having `-1` mean "keep expanding" is that we can keep subtracting from it and stay in the `depth<0` keep expanding behaviour.

File: pennylane/ops/op_math/sprod.py
Comment (albi3ro) [inline pennylane/ops/op_math/sprod.py:216]: I think we can ignore `id` for the simplify process.  I'm not actually quite sure what the use cases for `id` are. But currently in PennyLane, expansion and transformation processes tend to ignore it.  I think simplify should have that same behavior.

Comment (josh146) [reply pennylane/ops/op_math/sprod.py:216]: The `id` is used by developers/users who want to name a specific operation, and track it later in the process. It is used quite a bit by the Fourier module.

However, we never really worked out what makes the most sense if an operation with id is decomposed

File: pennylane/ops/qubit/hamiltonian.py
Comment (albi3ro) [inline pennylane/ops/qubit/hamiltonian.py:345]: Maybe instead we could do something like:

```suggestion
    def simplify(self, depth=None):
    
    if depth is not None:
        return self
```

That way we can still call this method in the conventional way, but we can also handle something like:

```
sum(qml.PauliX(0), H).simplify()
```

File: pennylane/ops/functions/equal.py
Comment (AlbertMitjans) [reply pennylane/ops/functions/equal.py]: Yep, you are right. I added the changes here because I needed the `arithmetic_depth` property. I created a new [PR](https://github.com/PennyLaneAI/pennylane/pull/2877) that will be merged after this one.

**Comment by AlbertMitjans** [comment]
> I think its worth thinking about whether we want `depth` to mean:
> 
> 1. Number of times we run the the `simplify` method
> 2. Maximum decrease in `arithmetic_depth`
> 
> I think I prefer the first option. Then we would only decrease `depth` by one when simplifying the next layer of operations.
> 
> If we only decrease by one, we don't have to worry about skipping from `1` to `-1`, bypassing zero.

If we go with the first option, what happens when simplifying a `Sum` operator? If we have `depth=1` then only the first summand is simplified?

File: pennylane/ops/op_math/sprod.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/sprod.py]: `id` ignored! 👍

File: pennylane/ops/qubit/hamiltonian.py
Comment (AlbertMitjans) [reply pennylane/ops/qubit/hamiltonian.py:345]: Do we want to call this `simplify` method when simplifying nested operators? Soran mentioned that this method is super ineffective.

**Note:** Calling `sum(qml.PauliX(0), H).simplify()` won't crash becaus the `simplify` method of the sum calls `H.simplify(depth=depth)`, which corresponds to the `simplify` method defined in the `Operator` class.

File: pennylane/ops/op_math/adjoint_class.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/adjoint_class.py:245]: I don't think we can remove this check right now. The `depth > 1` is needed because the depth is reduced by 2 (unless we change the meaning of the `depth` argument). If we remove the `depth < 0` check, then:
```python
qml.adjoint(qml.adjoint(qml.PauliX(0))).simplify() = qml.adjoint(qml.adjoint(qml.PauliX(0)))
```

**Comment by albi3ro** [comment]
`depth=1` would show the same behaviour for both options.

It would differ for something like `Adjoint(Adjoint(op))` and `depth=2`.

Option 1 would run simplify on `op`
Option 2 would stop and not simplify `op`

File: pennylane/ops/op_math/adjoint_class.py
Comment (albi3ro) [inline pennylane/ops/op_math/adjoint_class.py:241]: What we might want to do is specify `do_queue=False` whenever an operation is constructed within `simplify`.  Then, we manually queue the result in `qml.simplify` via `op.queue()`.

We could also just ignore `do_queue` in the `op.simplify` method and use `qml.tape.stop_recording()` in the `qml.simplify` method.

File: pennylane/ops/op_math/controlled_class.py
Comment (albi3ro) [inline pennylane/ops/op_math/controlled_class.py:316]:
```suggestion
        return Controlled(
```

`ControlledOp` is just for base ops that inherit from `Operation`, not just `Operator`.  We want to get rid of the distinction eventually, but for now, it's still there.

For example, what if we had a controlled projector? Or a controlled Hamiltonian?

File: pennylane/ops/op_math/pow_class.py
Comment (albi3ro) [inline pennylane/ops/op_math/pow_class.py:255]:
```suggestion
        if isinstance(self.base, Controlled):  # Pow(Controlled(base)) = Controlled(Pow(base))
```

File: pennylane/ops/op_math/pow_class.py
Comment (albi3ro) [inline pennylane/ops/op_math/pow_class.py:256]:
```suggestion
            return Controlled(
```

File: pennylane/ops/op_math/pow_class.py
Comment (Jaybsoni) [reply pennylane/ops/op_math/pow_class.py:257]: This example brings up a good question, suppose a certain simplification doesn't change the `depth`, (eg. a*(b+c) --> a*b + a*c). In this case, how can we make it clear to the user that simplification has occurred? Is there a way to quantify these steps so that we can give the user control over how many of these calls to make?

File: pennylane/ops/op_math/sum.py
Comment (Jaybsoni) [inline pennylane/ops/op_math/sum.py:319]: I don't think we need the continue here?

**Review by Jaybsoni — COMMENTED** [review]
Looks good! Added a few comments. Waiting for the changes to Prod to be in before I give the approve 💯

File: pennylane/ops/op_math/sum.py
Comment (soranjh) [reply pennylane/ops/op_math/sum.py:302]: We might reserve `simplify` for the same action that the current `simplify` method does.

File: pennylane/ops/op_math/adjoint_class.py
Comment (albi3ro) [inline pennylane/ops/op_math/adjoint_class.py:246]:
```suggestion
        if isinstance(self.base, Adjoint):  # Adj(Adj(A)) = A
            return self.base.base.simplify(depth=(0 if depth==1 else depth-2))
```

File: pennylane/ops/op_math/controlled_class.py
Comment (albi3ro) [inline pennylane/ops/op_math/controlled_class.py:316]:
```suggestion
        if isinstance(self.base, Controlled):
            base = self.base.base.simplify(depth=0 if depth==1 else depth-2)
            return Controlled(base,
                control_wires = self.control_wires + self.base.control_wires,
                control_values = self.control_values + self.base.control_values,
                work_wires = self.work_wires + self.base.work_wires
            )
        return ControlledOp(
```
This would cover:
```
op1 = Controlled(qml.PauliX(0), 1)
op2 = Controlled(op1, 2)
op2.simplify()
```

File: pennylane/ops/op_math/sprod.py
Comment (albi3ro) [reply pennylane/ops/op_math/sprod.py:203]: +1

File: pennylane/ops/op_math/sum.py
Comment (albi3ro) [reply pennylane/ops/op_math/sum.py:302]: I'm fine with the name `simplify_summands`.

But I do think we should make it a private method `_simplify_summands`.  That gives us more flexibility to change it in the future, and it should only ever be accessed from `op.simplify` anyway.

File: pennylane/ops/qubit/hamiltonian.py
Comment (soranjh) [reply pennylane/ops/qubit/hamiltonian.py:345]: > Soran mentioned that this method is super inefficient.

There is a version of [simplify](https://github.com/PennyLaneAI/pennylane/blob/c868132deef9c474a3422f1ca485251a29db1be0/pennylane/qchem/observable_hf.py#L200) inside qchem that is more efficient.

Comment (albi3ro) [reply pennylane/ops/qubit/hamiltonian.py:345]:
```
(1.0 * qml.PauliX(0)).simplify(depth=2)
```
Does crash.

We want to make it so that if a `Hamiltonian` is called without a depth keyword argument, like `H.simplify()`, it does the expected simplification in-place. But, if it's called *with* a depth keyword argument, then we just `return self`.

Comment (soranjh) [reply pennylane/ops/qubit/hamiltonian.py:345]: We should have a unified logic and only one `simplify` function though.

File: pennylane/ops/op_math/sprod.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/sprod.py]: Moved `arithmetic_depth` to `SymbolicOp` class in commit d22ba7fd15f55cfb3144f988071f9b67ef80abc9.

File: pennylane/ops/op_math/sum.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/sum.py]: Changed the method name to `_simplify_summands`. I also changed it to a `@classmethod` to avoid a `protected-access` pylint error. (see commit b1787b6eb7fb8b1a58bc7db971571c442de96a20)

File: pennylane/ops/op_math/sum.py
Comment (Jaybsoni) [reply pennylane/ops/op_math/sum.py:330]: Ahh ok I see what's happening. 💯

File: pennylane/ops/op_math/adjoint_class.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/adjoint_class.py]: Depth has been removed from simplify.

File: pennylane/ops/qubit/hamiltonian.py
Comment (AlbertMitjans) [reply pennylane/ops/qubit/hamiltonian.py:345]: Depth has been removed from `simplify` method.

File: pennylane/ops/op_math/pow_class.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/pow_class.py:251]: We removed the `depth` argument from the `simplify` method. We can add a `max_recursion` argument if we see that simplifying becomes computationally expensive.

Comment (AlbertMitjans) [reply pennylane/ops/op_math/pow_class.py:251]: And yes @albi3ro ! I forgot to simplify `self.base.base`.

File: pennylane/ops/op_math/adjoint_class.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/adjoint_class.py]: Depth has been removed.

File: pennylane/ops/op_math/prod.py
Comment (albi3ro) [inline pennylane/ops/op_math/prod.py:334]:
```suggestion
```

File: pennylane/ops/qubit/hamiltonian.py
Comment (albi3ro) [inline pennylane/ops/qubit/hamiltonian.py:345]:
```suggestion
    def simplify(self):
```

File: pennylane/ops/op_math/controlled_class.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/controlled_class.py]: Nice catch! Added.

File: pennylane/ops/op_math/prod.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/prod.py]: Done!

File: pennylane/ops/op_math/adjoint_class.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/adjoint_class.py:237]: Change added in PR https://github.com/PennyLaneAI/pennylane/pull/2854

File: doc/releases/changelog-dev.md
Comment (Jaybsoni) [inline doc/releases/changelog-dev.md:399]: This is from the other PR I believe? maybe we can remove it here and just pull master once it gets merged

Comment (Jaybsoni) [reply doc/releases/changelog-dev.md:399]: Or was it required to support some simplify logic?

File: pennylane/operation.py
Comment (Jaybsoni) [inline pennylane/operation.py:1215]: Still a bit unclear what this method is doing, does it simplify to `arithmetic_depth` or, does it maximally simplify? 
```suggestion
        """Reduces to the maximum depth of nested operators.
```

File: pennylane/ops/op_math/adjoint_class.py
Comment (Jaybsoni) [inline pennylane/ops/op_math/adjoint_class.py:249]: 💯

File: pennylane/ops/op_math/pow_class.py
Comment (Jaybsoni) [inline pennylane/ops/op_math/pow_class.py:248]: What if the base acts on multiple wires? Not sure if Pow supports that right now. This should be a Prod of Identities on each wire?

File: pennylane/ops/op_math/prod.py
Comment (Jaybsoni) [inline pennylane/ops/op_math/prod.py:347]: It seems problematic to me that these two cases are treated the same? Also codecov complaining about missing test coverage.

File: pennylane/ops/qubit/hamiltonian.py
Comment (Jaybsoni) [inline pennylane/ops/qubit/hamiltonian.py:378]: 🥲

File: pennylane/ops/op_math/prod.py
Comment (Jaybsoni) [reply pennylane/ops/op_math/prod.py]: Would the following two input factors yield the same output? 

- `factors = ( X, Sum(Y, Z) )`. ---> `new_factors = (X, Y, Z)`
- `factors = ( X, Adjoint(Prod(Z, Y)) )` ---> `new_factors = (X, Y, Z)`

File: doc/releases/changelog-dev.md
Comment (AlbertMitjans) [reply doc/releases/changelog-dev.md]: I wanted to use the dunder methods in some tests, that's why I merged it. Now the PR is merged into master, so this change shouldn't appear.

File: pennylane/ops/op_math/prod.py
Comment (albi3ro) [inline pennylane/ops/op_math/prod.py:350]:
```suggestion
            elif not isinstance(simplified_factor, qml.Identity):
                new_factors += ((simplified_factor,),)
```
Or am I missing something?

File: pennylane/operation.py
Comment (AlbertMitjans) [reply pennylane/operation.py:1215]: It maximally simplifies. I will change the docstring to make it clearer.

File: pennylane/ops/op_math/pow_class.py
Comment (albi3ro) [reply pennylane/ops/op_math/pow_class.py:248]: I think `Identity(wires=self.wires[0])` is fine. We just need to have something that's "an operation but does nothing".

File: pennylane/operation.py
Comment (AlbertMitjans) [reply pennylane/operation.py]: Hopefully it is clearer now.

File: pennylane/ops/op_math/pow_class.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/pow_class.py]: Done

File: pennylane/ops/op_math/prod.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/prod.py]: I just changed this XD. They have different behaviour now.

File: pennylane/ops/qubit/hamiltonian.py
Comment (AlbertMitjans) [reply pennylane/ops/qubit/hamiltonian.py:378]: [Sourcery](https://sourcery.ai/) is amazing... 🙄

File: pennylane/ops/op_math/prod.py
Comment (AlbertMitjans) [reply pennylane/ops/op_math/prod.py:350]: You are not missing anything 😓

File: pennylane/ops/qubit/hamiltonian.py
Comment (albi3ro) [reply pennylane/ops/qubit/hamiltonian.py:378]: I'll have to check this out for myself

**Review by Jaybsoni — APPROVED** [review]
Awesome! Looks great!

**Review by albi3ro — APPROVED** [review]
A lot more in this PR than I thought we'd get done! Excellent job 🎉

**Pull request merged**
